xdem/spatial_tools.py

```diff
@@ -398,6 +408,149 @@ def hillshade(dem: Union[np.ndarray, np.ma.masked_array], resolution: Union[floa
+    """
+
+    # get percentage of spread from the minimal cost
+    ind_min = cost.argmin()
+    min_cost = cost[ind_min]
+    perc_cost_improv = (cost - min_cost) / min_cost
+
+    # costs below threshold and lesser degrees
+    below_margin = np.logical_and(perc_cost_improv < margin_improvement / 100., np.arange(len(cost))<=ind_min)
+    costs_below_thresh = cost[below_margin]
+    # minimal costs
+    subindex = costs_below_thresh.argmin()
+    # corresponding index (degree)
+    ind = np.arange(len(cost))[below_margin][subindex]
+
+    if verbose:
+        print('Polynomial of degree '+str(ind_min+1)+ ' has the minimum cost value of '+str(min_cost))
+        print('Polynomial of lesser degree '+str(ind+1)+ ' is selected within a '+str(margin_improvement)+' % margin of'
+            ' the minimum cost, with a cost value of '+str(min_cost))
+
+    return ind
+
+
+def robust_polynomial_fit(x: np.ndarray, y: np.ndarray, max_order: int = 6, estimator: str  = 'Theil-Sen',
+                          cost_func: Callable = median_absolute_error, margin_improvement : float = 20.,
+                          linear_pkg = 'sklearn', verbose: bool = False, random_state = None, **kwargs) -> tuple[np.ndarray,int]:
+    """
+    Given sample 1D data x, y, compute a robust polynomial fit to the data. Order is chosen automatically by comparing
+    residuals for multiple fit orders of a given estimator.
+    :param x: input x data (N,)
+    :param y: input y data (N,)
+    :param max_order: maximum polynomial order tried for the fit
+    :param estimator: robust estimator to use, one of 'Linear', 'Theil-Sen', 'RANSAC' or 'Huber'
+    :param cost_func: cost function taking as input two vectors y (true y), y' (predicted y) of same length
+    :param margin_improvement: improvement margin (percentage) below which the lesser degree polynomial is kept
+    :param linear_pkg: package to use for Linear estimator, one of 'scipy' and 'sklearn'
+    :param random_state: random seed for testing purposes
+    :param verbose: if text should be printed
+
+    :returns coefs, degree: polynomial coefficients and degree for the best-fit polynomial
+    """
+    if not isinstance(estimator, str) or estimator not in ['Linear','Theil-Sen','RANSAC','Huber']:
+        raise ValueError('Attribute estimator must be one of "Linear", "Theil-Sen", "RANSAC" or "Huber".')
+    if not isinstance(linear_pkg, str) or linear_pkg not in ['sklearn','scipy']:
+        raise ValueError('Attribute linear_pkg must be one of "scipy" or "sklearn".')
+
+    # select sklearn estimator
+    dict_estimators = {'Linear': LinearRegression(), 'Theil-Sen':TheilSenRegressor(random_state=random_state)
+        , 'RANSAC': RANSACRegressor(random_state=random_state), 'Huber': HuberRegressor()}
+    est = dict_estimators[estimator]
+
+    # TODO: this should be a function outside (waiting for Amaury's PR)
+    # remove NaNs
+    mykeep = np.logical_and.reduce((np.isfinite(y), np.isfinite(x)))
```

Comment:
What does `np.logical_and.reduce` do? I've never seen it before!

Reply:
Here it is useless (again, remnant of a previous version of the code).

Basically, it allows you to concatenate as many conditions as you desire:
`np.logical_and.reduce((cond1,cond2,cond3))` is equivalent to `np.logical_and(np.logical_and(cond1,cond2),cond2)` 
Quite useful :wink:

Reply:
Aah, nice! Thanks!!